sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala

```diff
@@ -1393,4 +1401,58 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
       )
     }
   }
+
+  test("SPARK-11544 test pathfilter") {
+    def filterFunc(file: File): Boolean = {
+      val name = file.getCanonicalFile.getName()
+      val ret = !(name.startsWith(".")  || name.startsWith("_"))
+      ret.asInstanceOf[Boolean]
+    }
+
+    val clonedConf = new Configuration(hadoopConfiguration)
+    try {
+      val dir = Utils.createTempDir()
+      dir.delete()
+      val path = dir.getCanonicalPath
+      val emps = Seq(
+        Row(1, "Emp-1"),
+        Row(2, "Emp-2")
+      )
+
+      val empRows = sqlContext.sparkContext.parallelize(emps, 1)
+      val empSchema = StructType(
+        Seq(
+          StructField("id", IntegerType, true),
+          StructField("name", StringType, true)
+        )
+      )
+      val empDF = sqlContext.createDataFrame(empRows, empSchema)
+      empDF.write.json(path)
+      val files = dir.listFiles().filter(filterFunc(_))
+      files.map(_.renameTo(new File(path + File.separator + "pathmap.tmp")))
+      FileUtils.copyFile(new File(path + File.separator + "pathmap.tmp"),
+        new File(path + File.separator + "data.json"))
+
+      // Both the files should be read and count should be 4
+      val empDF2 = sqlContext.read.json(path)
+      empDF2.registerTempTable("jsonTable1")
+      checkAnswer(
+        sql("select count(*) from jsonTable1"),
+        Row(4)
+      )
+      // After applying the path filter, only one file should be read and the count should be 2
+      sqlContext.sparkContext.hadoopConfiguration.setClass("mapreduce.input.pathFilter.class",
+        classOf[TmpFileFilter], classOf[PathFilter])
+      val empDF3 = sqlContext.read.json(path)
+      empDF3.registerTempTable("jsonTable2")
+      checkAnswer(
+        sql("select count(*) from jsonTable2"),
+        Row(2)
+      )
+    } finally {
+      // Hadoop 1 doesn't have `Configuration.unset`
+      hadoopConfiguration.clear()
+      clonedConf.asScala.foreach(entry => hadoopConfiguration.set(entry.getKey, entry.getValue))
+    }
+  }
```

Comment:
@liancheng wow.. thanks a lot. This indeed looks much better. Have implemented the all other comments and have re-pushed.